Homework #6. Task #1. Волынец Арсений

vol0n opened the pull request.

vol0n committed:
Initial commit

vol0n committed:
Make plots with lets-plot

elena-lyulina commented on build.gradle.kts:
```diff
@@ -25,6 +25,13 @@ dependencies {
+    // for plotting with lets-plot
+    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
+    implementation( "org.jetbrains.lets-plot:lets-plot-jfx:2.0.1")
+    api( "org.jetbrains.lets-plot:lets-plot-common:2.0.1")
+    api( "org.jetbrains.lets-plot:lets-plot-kotlin-api:2.0.1")
```
ты уверен, что тебе здесь api зависимость нужна?

vol0n replied:
Действительно, одну из них можно убрать, а вот lets-plot-kotlin-api нет.

elena-lyulina commented on src/main/kotlin/homework6/buildGraph.kt:
```diff
@@ -0,0 +1,43 @@
+
```
файлы называем по стайлгайду https://kotlinlang.org/docs/coding-conventions.html#source-file-names

еще я посмотрела, у тебя в других домашках тоже не по стайлгайду называются файлы типа task_i.kt, так не должно быть, давай ты в отдельной ветке их поправишь и сделаешь на меня пр

elena-lyulina commented on src/main/kotlin/homework6/buildGraph.kt:
```diff
@@ -0,0 +1,43 @@
+import kotlin.math.round
+
+fun buildGraph() {
+    val tst = Json.decodeFromString<GlobalTest>(File("results.json").readText())
```
а что с названиями переменных произошло? почему они внезапно состоят из одной буквы? давай ты что-то более осмысленное будешь использовать, это не питон :)

elena-lyulina commented on src/main/kotlin/homework6/buildGraph.kt:
```diff
@@ -0,0 +1,43 @@
+    var plotPos = 0
+    @Suppress("MagicNumber")
+    for (test in tst.tests) {
+        val curp = p +
```
что за курп?

elena-lyulina commented on src/main/kotlin/homework6/buildGraph.kt:
```diff
@@ -0,0 +1,43 @@
+    val p = letsPlot(d) { x = "thrNo" }
+    val bunch = GGBunch()
+    var plotPos = 0
+    @Suppress("MagicNumber")
```
ух, не надо просто посередине кода использовать supress, тем более тут он вполне осмысленно ругается
я вот тоже например не понимаю, откуда у тебя взялись числа типа 400 
давай ты их в константы что ли вынесешь с осмысленными именами какими-то

elena-lyulina commented on src/main/kotlin/homework6/buildGraph.kt:
```diff
@@ -0,0 +1,43 @@
+import kotlin.math.log
+import kotlin.math.round
+
+fun buildGraph() {
```
давай это все же будет классом с параметрами типа "откуда брать данные" и "куда сохранять файл", а то у тебя пути с неба откуда-то свалились

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -0,0 +1,85 @@
+data class TestData(val threadsNums: List<Int>, val arraySizes: List<Int>)
+
+// data[l, m], data[m+1, r] are sorted, merge them into sorted data[l, r]
+fun merge(data: IntArray, l: Int, m: Int, r: Int) {
```
опять же с названиями беда, ничего не понятно, давай какие-то более осмысленные и желательно из одного-двух слов

elena-lyulina replied:
а еще merge тоже хочется многопоточным иметь

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -0,0 +1,85 @@
+
+// data[l, m], data[m+1, r] are sorted, merge them into sorted data[l, r]
+fun merge(data: IntArray, l: Int, m: Int, r: Int) {
+    val left = IntArray(m - l + 1) { i -> data[i + l] }
+    val right = IntArray(r - m) { i -> data[i + m + 1] }
```
можешь вместо i использовать it внутри лямбды, так по дефолту можно обращаться к параметру, а конкретно здесь вообще лучше использовать sliceArray, например

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -0,0 +1,85 @@
+        if (numberOfThreads != 1) {
+            val th = Thread { sort(data, l, mid, numberOfThreads / 2) }
+            th.start()
+            sort(data, mid + 1, r, numberOfThreads - numberOfThreads / 2)
```
а это почему не в отдельном потоке запускается?

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -0,0 +1,85 @@
+    }
+}
+
+fun test(testName: String) {
```
давай вот это вынесем куда-то от реализации сортировки, отделяй показательные запуски от реализации, вдруг реализацией потом еще кто-то будет пользоваться, зачем ему тут твои запуски

elena-lyulina replied:
ну и название странное, тесты это все же то, что в папке тестов лежит, а тут ты время замеряешь и все такое

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -0,0 +1,85 @@
+    val testData = Json.decodeFromString<TestData>(object {}.javaClass.getResource(testName).readText())
+    val rand = java.util.Random()
+
+    val file = File("results.json")
```
название файлов тоже лучше в куда-то в переменную вынести

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -0,0 +1,85 @@
+import kotlinx.serialization.json.Json
+import java.io.File
+
+@Serializable
+data class TestPerThread(val arraySize: Int, val timeArray: List<Double>)
+
+@Serializable
+data class GlobalTest(val threadsNums: List<Int>, val tests: List<TestPerThread>)
+
+@Serializable
+data class TestData(val threadsNums: List<Int>, val arraySizes: List<Int>)
```
вот эти названия тоже не очень показательные
что такое глобал тест? что такое тест дата? надо какие-то более специфичные названия, отражающие суть того, что ты там хранишь

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -0,0 +1,85 @@
+            startTime = System.nanoTime()
+            sort(test, 0, test.lastIndex, threadNum)
+            stopTime = System.nanoTime()
+            @Suppress("MagicNumber")
```
давай все же ты будешь прислушиваться к тому, что тебе пишет детект, использование магических чисел действительно так себе и не должно быть повсеместно
в данном случае можно было например поискать, как конвертить в секунды какими-то встроенными методами, а не собственными силами, так как это всегда более надежно
например, есть метод TimeUnit.NANOSECONDS.toSeconds

elena-lyulina commented on src/test/kotlin/homework6/ParallelMergeSortKtTest.kt:
```diff
@@ -0,0 +1,42 @@
+    @MethodSource("inputDataSort")
+    @ParameterizedTest
+    fun testSortMultiThread(actual: IntArray, expected: IntArray) {
+        sort(actual, 0, actual.lastIndex, 2)
+        assertArrayEquals(expected, actual)
+        sort(actual, 0, actual.lastIndex, 5)
+        assertArrayEquals(expected, actual)
+        sort(actual, 0, actual.lastIndex, 10)
+        assertArrayEquals(expected, actual)
```
давай потоки тоже будут параметром, делать несколько ассертов в одном тесте -- так себе идея

elena-lyulina commented on src/test/kotlin/homework6/ParallelMergeSortKtTest.kt:
```diff
@@ -0,0 +1,42 @@
+import org.junit.jupiter.params.provider.Arguments
+import org.junit.jupiter.params.provider.MethodSource
+
+fun IntArray.print() {
+    this.forEach { x -> print("$x ") }
+    println()
+}
```
ты это нигде не используешь, кажется

elena-lyulina reviewed (changes requested):
Хочется иметь возможность легко сравнивать, как ведут себя сортировки с разным количеством потоков. При помощи твоих диаграмм это конечно можно сделать, если нарисовать их рядом, но это все еще не самый оптимальный выбор графа

я бы сказала, что хочется видеть что-то вроде такого (или любой другой тип графиков, на которых реально было бы удобно сравнивать различные сортировки)
![image](https://user-images.githubusercontent.com/24303332/116878340-e0ca7a00-ac27-11eb-9478-bd06b7b07ca6.png)

vol0n committed:
Remove lets-plot-common api dependence

vol0n committed:
Required changes:

- make merge multithreaded
- change type of plot to pathplot
- separate measuring performance launches and sort implementation
- use measured time as a table (dataFrame)

elena-lyulina commented on src/main/kotlin/homework6/DoSortBenchmark.kt:
```diff
@@ -0,0 +1,97 @@
+package homework6
```
а почему название файла начинается с глагола-то?

elena-lyulina commented on src/main/kotlin/homework6/DoSortBenchmark.kt:
```diff
@@ -0,0 +1,97 @@
+        var i = 0
+        for (size in input.arraySizes) {
+            for (threadNo in input.threadsNums) {
+                val test = IntArray(size) { _ -> rand.nextInt() }
```
`_ ->` можно не писать внутри лямбды, если ты не используешь параметр

elena-lyulina commented on src/main/kotlin/homework6/DoSortBenchmark.kt:
```diff
@@ -0,0 +1,97 @@
+    private data class DataFrame(val threadNum: List<Int>, val arraySize: List<Int>, val time: List<Double>)
+    private var objForSerialization: DataFrame
+
+    companion object {
```
посмотри еще на порядок элементов в классе в стайлгайде, а то у тебя тут путаются свойства и companion object, сложно читать такой код

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -1,85 +1,110 @@
-
-@Serializable
-data class GlobalTest(val threadsNums: List<Int>, val tests: List<TestPerThread>)
+class ParallelMergeSort(val destinationArray: IntArray, val sourceArray: IntArray, initialNumberOfThreads: Int) {
```
свойства из констурктора можно сделать приватными

elena-lyulina replied:
и прислушивайся вообще, что тебе идея пишет)
![image](https://user-images.githubusercontent.com/24303332/119806324-c748e500-beea-11eb-8271-f77cbeece1e0.png)
![image](https://user-images.githubusercontent.com/24303332/119806348-cb750280-beea-11eb-8dda-b3c68319b656.png)

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -1,85 +1,110 @@
+        val size: Int
+            get() = right - left + 1
+        val middle: Int
+            get() = (right + left) / 2
```
в такой реализации у тебя каждый раз при использовании свойства оно будет считаться заново, тебе это не нужно, можно сразу их инициализировать нужным значением, без переопределения get()

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -1,85 +1,110 @@
-        val mid = (l + r) / 2
+   // sort sourceArray[l, r] and save into destArray[s, s + r - l]
+   private fun sort(sa: SubArray, destArray: IntArray, s: Int, numberOfThreads: Int = 1) {
+        if (sourceArray.isEmpty()) return
+        if (sa.size == 1) {
+            destArray[s] = sourceArray[sa.left]
+            return
+        }
+        val tempArray = IntArray(sa.size)
+        // number of elements in the right part of subArray sa
+        val n1 = sa.middle - sa.left + 1
```
извини, но я все равно буду придираться к именам переменных
из [naming conventions](https://kotlinlang.org/docs/coding-conventions.html#choose-good-names):

> The names should make it clear what the purpose of the entity is, so it's best to avoid using meaningless words


читатель твоего кода в идеале должен понимать, в чем смысл переменной, не обращаясь к ее инициализации (и особенно забавно, что в одном месте ты написал коммент, что это за переменная такая, когда это все должно быть понятно из имени). поэтому sa, n1, th, s, l, r и все твои остальные имена надо сделать осмысленными! используй слова целиком, используй несколько слов вместе, код должен быть понятным и в перспективе поддерживаемым долгие годы разными программистами

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -1,85 +1,110 @@
-
-@Serializable
-data class GlobalTest(val threadsNums: List<Int>, val tests: List<TestPerThread>)
+class ParallelMergeSort(val destinationArray: IntArray, val sourceArray: IntArray, initialNumberOfThreads: Int) {
```
еще странно выглядит, что надо передавать destination array, это не очень удобно для использования
если тебе нужен какой-то доп массив, можно создавать его уже внутри класса или сразу менять передаваемый массив, а то у тебя все вызовы выглядят вот так ParallelMergeSort(array, array, ...)

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -1,85 +1,110 @@
-data class GlobalTest(val threadsNums: List<Int>, val tests: List<TestPerThread>)
+class ParallelMergeSort(val destinationArray: IntArray, val sourceArray: IntArray, initialNumberOfThreads: Int) {
+    init {
+        sort(SubArray(0, sourceArray.lastIndex), destinationArray, 0, initialNumberOfThreads)
```
по логике чуть странно, что ты создаешь объект класса и он сразу сортирует твой массив, если ты посмотришь на встроенные сортировки в котлине, то это все же функции, куда передается массив или которые расширяют массив. 
можно сделать из класса объект, в котором будет одна публичная функция sort и ее уже вызывать

elena-lyulina replied:
то, что ты сделал отдельную функцию, которая создает экземпляр класса и вызывает у него sort(), не решает особо проблемы, более того это очень странно выглядит
вопрос в том, зачем вообще создавать экземпляр класса чтобы отсортировать массив, хочется передавать все параметры напрямую в функцию. тем более что ты ни разу не вызываешь класс с sortSourceArray: Boolean = false, то есть ты всегда хочешь сортировать сразу исходный массив, поэтому хранить свойство destinationArray не очень-то и надо
можно сразу в функции sort создавать временный массив и потом копировать его значение в исходный массив

и наконец сделать из класса объект

elena-lyulina reviewed (changes requested):
уже близко! однако, график все еще не очень показательный
почему у тебя есть какие-то скачки? почему соединены точки только для одного потока? по нему все еще непонятно, какие сортировки работают лучше, а какие хуже, мб ты какие-то не те точки просто соединяешь?
![image](https://user-images.githubusercontent.com/24303332/119809311-a59d2d00-beed-11eb-9ff1-3b6a16b436df.png)

vol0n committed:
Required changes:

- Fix names
- Fix order entities in class body
- Fix plot generation
- Use sortMT, instead of creating a class and calling its method

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -0,0 +1,133 @@
+        val tempArray = IntArray(sourceArray.size)
+        // number of elements in the right part of subArray sa
+        val numOfElemsInRightPartOfSourceArray = sourceArray.middle - sourceArray.left + 1
+        if (numberOfThreads != 1) {
```
а если numberOfThreads < 1?

elena-lyulina reviewed (changes requested):
поправь еще, пожалуйста, один из предыдущих комментариев

vol0n committed:
Make ParallelMergeSort object, add error

- ParallelMergeSort is now an object, not a class.
- An exception is thrown when initialNumberOfThreads < 1.

vol0n committed:
Change reflect dependency version

vol0n committed:
Resolve conflicts

vol0n committed:
Fix deprecated usage: inSeconds

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -1,25 +1,28 @@
-    sortSourceArray: Boolean = false
-) {
-    private var destinationArray = if (sortSourceArray) sourceArray else IntArray(sourceArray.size)
+fun sortMT(sourceArray: IntArray, numberOfThreads: Int) =
+    ParallelMergeSort.sortMT(sourceArray, numberOfThreads)
```
вот этой функции вообще не надо, зачем она нужна? ты же можешь сразу вызывать сорт у объекта везде

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -1,25 +1,28 @@
+fun sortMT(sourceArray: IntArray, numberOfThreads: Int) =
+    ParallelMergeSort.sortMT(sourceArray, numberOfThreads)
 
+private object ParallelMergeSort {
```
почему он стал приватным?.......... 
все нормально было же с модификаторами

то, что стал объектом, правильно

elena-lyulina reviewed (changes requested):
почтиии

vol0n committed:
Required changes and add interface Sorter

- remove sortMT global function
- add Interface sorter

vol0n committed:
Fix tests

elena-lyulina commented on src/main/kotlin/homework6/ParallelMergeSort.kt:
```diff
@@ -1,27 +1,35 @@
 
-fun sortMT(sourceArray: IntArray, numberOfThreads: Int) =
-    ParallelMergeSort.sortMT(sourceArray, numberOfThreads)
+enum class ResourceKind(val kind: String) {
+    THREADS("Threads"), COROUTINES("Coroutines")
+}
```
вообще это не должно было появиться в этом пр, но раз до зачета один час, то ладно

elena-lyulina reviewed (approved):
поправь конфликты и солью)

vol0n committed:
Fix conflicts
